interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -0,0 +1,245 @@
+package io.grpc.testing.integration;
+
+import io.netty.channel.socket.nio.NioSocketChannel;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+// import java.lang.annotation.Native;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketException;
+import java.net.SocketImpl;
+import java.util.ArrayList;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.logging.Logger;
+
+public class TrafficControlProxy {
+
+  private static int DEFAULT_BAND = 1 * 1024 * 1024; // Bps
+  private static int DEFAULT_DELAY_MS = 100; // ms;
+  private static Logger logger = Logger.getLogger(TrafficControlProxy.class.getName());
+
+  private String loopback = "127.127.127.127";
+  // TODO: make ports arguments
+  private int proxyPort = 5050;
+  private int serverPort = 5001;
+  private int qLength;
+  private int serverQSize;
+  private int clientQSize;
+  private int chunkSize;
+  private boolean shutDown;
+  private LinkedBlockingQueue<Message> serverQueue = new LinkedBlockingQueue<Message>();
+  private LinkedBlockingQueue<Message> clientQueue = new LinkedBlockingQueue<Message>();
+  private ServerSocket client;
+  private Socket server;
+  private Socket clientSock;
+
+
+  private int bandwidth = DEFAULT_BAND;
+  private int latency = DEFAULT_DELAY_MS;
+
+  public void setBandwidth(int targetBPS) {
+    bandwidth = targetBPS;
+  }
+
+
+  public void setLatency(int targetMillis) {
+    latency = targetMillis / 2; // divide by two because delay is applied in both directions
+  }
+
+  public void setUp() throws IOException {
+    qLength = Math.max(bandwidth * latency / 1000, 1); // convert from ms to s
+    chunkSize = Math.min(qLength, 4 * 1024);
+
+    logger.info("Started new proxy on port " + proxyPort);
+    client = new ServerSocket();
+    client.bind(new InetSocketAddress(loopback, proxyPort));
+    clientSock = client.accept();
+    server = new Socket();
+    server.connect(new InetSocketAddress(loopback, serverPort));
```

Comment:
terminology is going to be important here.  The "server" could be the proxy itself, since it accepts connections from other endpoints.  The server could also mean the destination that this proxy forwards to.